File: packages/sdk/src/modules/auditors.ts
import { AuditReportStruct } from "../artifacts/typechain/src/ConfidentialTransfers.js"
import { Payload } from "../index.js"
import { Inputs } from "./inputs.js"

export class Auditors extends Inputs {
  async createStateAuditReport(
    cPrivateKey: bigint,
    nonce: bigint,
    auditorAddresses: string[],
  ): Promise<AuditReportStruct[]> {
    const otk = await this.generateOTK(cPrivateKey, nonce)
    return await this.createAuditReport(
      cPrivateKey,
      otk,
      nonce,
      auditorAddresses,
    )
  }

  async createTransferAuditReport(
    cPrivateKey: bigint,
    nonce: bigint,
    recipientAddress: string,
    auditorAddresses: string[],
  ): Promise<AuditReportStruct[]> {
    const { pubKeyX, pubKeyY } = await this.token.getAccount(recipientAddress)

    const sharedKey = await Inputs.deriveSharedKey(
      cPrivateKey,
      pubKeyX,
      pubKeyY,
    )
    const otk = await this.generateOTK(sharedKey, nonce)
    return await this.createAuditReport(
      cPrivateKey,
      otk,
      nonce,
      auditorAddresses,
    )
  }

  private async createAuditReport(
    cPrivateKey: bigint,
    otk: bigint,
    nonce: bigint,
    auditorAddresses: string[],
  ): Promise<AuditReportStruct[]> {
    const { pubKeyXs, pubKeyYs } =
      await this.token.getCPublicKeys(auditorAddresses)
    const stateAuditReports: AuditReportStruct[] = []
    for (let i = 0; i < auditorAddresses.length; i++) {
      const auditorAddress = auditorAddresses[i]
      const sharedKey = await Inputs.deriveSharedKey(
        cPrivateKey,
        pubKeyXs[i],
        pubKeyYs[i],
      )
      const eOTK = await Inputs.cipher(sharedKey, nonce, otk)
      stateAuditReports.push({
        auditor: auditorAddress,
        eOTK: eOTK,
      })
    }
    return stateAuditReports
  }

  async decryptAuditReport(
    cPrivateKey: bigint,
    senderAddress: string,
    eOTK: bigint,
    payload: Payload,
  ): Promise<bigint> {
    const { pubKeyX, pubKeyY } = await this.token.getAccount(senderAddress)
    const sharedKey = await Inputs.deriveSharedKey(
      cPrivateKey,
      pubKeyX,
      pubKeyY,
    )
    const otk = await Inputs.decipher(
      sharedKey,
      BigInt(payload.nonce),
      BigInt(eOTK),
    )
    const amount = await Inputs.decipher(
      otk,
      BigInt(payload.nonce),
      BigInt(payload.eAmount),
    )

    const commitment = await Inputs.generateCommitment(amount, otk)

    if (commitment !== BigInt(payload.commitment))
      throw new Error("Commitment payload does not match")

    return amount
  }
}

File: packages/go-sdk/sdk/auditors.go
package sdk

import (
	"fmt"
	"math/big"

	"github.com/ethereum/go-ethereum/common"
	confidential_transfers "github.com/noxlabs/confidential-transfers-go-sdk/sdk/artifacts/contracts/ConfidentialTransfers"
)

func (s *SDK) CreateStateAuditReport(cPrivateKey *big.Int, nonce *big.Int, auditorAddresses []common.Address) ([]confidential_transfers.AuditReport, error) {
	otk, err := s.GenerateOTK(cPrivateKey, nonce)
	if err != nil {
		return nil, err
	}
	return s.createAuditReport(cPrivateKey, otk, nonce, auditorAddresses)
}

func (s *SDK) CreateTransferAuditReport(cPrivateKey *big.Int, nonce *big.Int, recipientAddress common.Address, auditorAddresses []common.Address) ([]confidential_transfers.AuditReport, error) {
	account, err := s.Token.GetAccount(nil, recipientAddress)
	if err != nil {
		return nil, err
	}

	sharedKey, err := DeriveSharedKey(cPrivateKey, account.PubKeyX, account.PubKeyY)
	if err != nil {
		return nil, err
	}

	otk, err := s.GenerateOTK(sharedKey, nonce)
	if err != nil {
		return nil, err
	}

	return s.createAuditReport(cPrivateKey, otk, nonce, auditorAddresses)
}

func (s *SDK) createAuditReport(cPrivateKey *big.Int, otk *big.Int, nonce *big.Int, auditorAddresses []common.Address) ([]confidential_transfers.AuditReport, error) {
	keys, err := s.Token.GetCPublicKeys(nil, auditorAddresses)
	if err != nil {
		return nil, err
	}

	pubKeyXs := keys.PubKeyXs
	pubKeyYs := keys.PubKeyYs

	var stateAuditReports []confidential_transfers.AuditReport

	for i, auditorAddress := range auditorAddresses {
		sharedKey, err := DeriveSharedKey(cPrivateKey, pubKeyXs[i], pubKeyYs[i])
		if err != nil {
			return nil, err
		}

		eOTK, err := Cipher(sharedKey, nonce, otk)
		if err != nil {
			return nil, err
		}

		stateAuditReports = append(stateAuditReports, confidential_transfers.AuditReport{
			Auditor: auditorAddress,
			EOTK:    eOTK,
		})
	}

	return stateAuditReports, nil
}

func (s *SDK) DecryptAuditReport(cPrivateKey *big.Int, senderAddress common.Address, eOTK *big.Int, payload confidential_transfers.Payload) (*big.Int, error) {
	account, err := s.Token.GetAccount(nil, senderAddress)
	if err != nil {
		return nil, err
	}

	sharedKey, err := DeriveSharedKey(cPrivateKey, account.PubKeyX, account.PubKeyY)
	if err != nil {
		return nil, err
	}

	otk, err := Decipher(sharedKey, payload.Nonce, eOTK)
	if err != nil {
		return nil, err
	}

	amount, err := Decipher(otk, payload.Nonce, payload.EAmount)
	if err != nil {
		return nil, err
	}

	commitment, err := GenerateCommitment(amount, otk)
	if err != nil {
		return nil, err
	}

	if commitment.Cmp(payload.Commitment) != 0 {
		return nil, fmt.Errorf("commitment payload does not match")
	}

	return amount, nil
}
